protocol/protocol.typeHierarchy.proposed.md

```diff
@@ -0,0 +1,77 @@
+
+#### Type Hierarchy
+
+Many language support inheritance and LSP provides retrieving the hierarchy information with the following requests.
+
+
+The `textDocument/subTypes` request is sent from the client to the server to resolve resolve the subtypes of a symbol at a given text document position. Returns the direct subtypes in no particular order.
+
+_Client Capabilities_:
+
+```ts
+  TextDocumentClientCapabilities {
+    /**
+     * Capabilities specific to the `textDocument/subTypes`
+     *  and `textDocument/superTypes
+     */
+    typeHierarchy?: {
+    /**
+     * Whether implementation supports dynamic registration.
+     */
+    dynamicRegistration?: boolean;
+  };
+```
+
+_Server Capabilities_:
+```ts
+ServerCapabilities{
+  /**
+   * The server provides type hierarchy information
+   */
+  typeHierarchyProvider?: boolean
+}
+```
+
+```ts
+/**
+ * Represents hierarchical information about programming constructs like variables, classes,
+ * interfaces etc.
+ */
+export interface SymbolNode extends SymbolInformation {
+  /**
+   * Immediate descendants of this node
+   */
+  descendants? : SymbolNode[]
+  /**
+   * true if this node has children. hasChildren can be
+   * true even when the descendants are empty. In that case
+   * the clients should do a call to retrieve the descendants.
+   */
+  hasChildren: boolean
+}
+```
+
+_Request_:
+
+method: ‘textDocument/subTypes’
+params: TextDocumentPositionParams
+
+_Response_
+
+* result: `SymbolNode[] | null`
+* error: code and message set in case an exception happens during the 'textDocument/subTypes' request
+
+
+
+The `textDocument/superTypes` request is sent from the client to the server to resolve resolve the supertypes of a symbol at a given text document position. Returns all the supertypes in bottom-up order.
+
+_Request_:
+
+method: ‘textDocument/superTypes’
+params: TextDocumentPositionParams
+
+_Response_
+
+* result: `SymbolNode[] | null`
```

Comment:
So a type can have multiple parent types? If that is true, how would it be displayed in a tree UI (vs a graph)?

Reply:
multiple-inheritance has been around and there have been tree based representations of it implemented before. The most generic approach I have seen is to replicate the same Type hierarchy tree under all parent nodes. We can add additional information to this data structure if it is going to help with its representation.

Reply:
I think you should reclarify the goal of this proposal, I misunderstood it. The linked issue describes a symbol hierarchy (e.g. namespaces -> classes -> members), not inheritance

Reply:
Also take a look at https://github.com/Microsoft/language-server-protocol/issues/472 I think that should include the symbol hirearchy